File: scripts/day2/day2_scripts_R/Day2_Application3_NIH.R
# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 2 - From raw output to usable data
# Application 3: Life Sciences - NIH RePORTER
# =============================================================================
# Goal: Clean and standardize raw JSON data into a usable tabular format.

library(jsonlite)
library(dplyr)
library(purrr)
library(fs)
library(tidyr)

clean_nih_data <- function() {
  raw_cache_file <- path(path_dir(path_dir(getwd())), "day1", "data", "cache", "day1_app3_nih_raw.json")
  clean_dir <- path(path_dir(path_dir(getwd())), "day2", "data", "cache")
  dir_create(clean_dir)
  clean_file <- path(clean_dir, "day2_app3_nih_clean.csv")
  
  cat("Loading raw data from", raw_cache_file, "\n")
  raw_data <- fromJSON(raw_cache_file, simplifyVector = FALSE)
  
  records <- raw_data$results
  cat("Loaded", length(records), "records. Starting cleaning process...\n")
  
  df <- map_dfr(records, function(rec) {
    org_name <- if (!is.null(rec$organization) && !is.null(rec$organization$org_name)) rec$organization$org_name else "Unknown"
    agency_name <- if (!is.null(rec$agency_ic_admin) && !is.null(rec$agency_ic_admin$name)) rec$agency_ic_admin$name else "Unknown"
    award_amt <- if (!is.null(rec$award_amount)) as.numeric(rec$award_amount) else 0
    
    tibble(
      ProjectNumber = rec$project_num %||% "",
      Title = rec$project_title %||% "",
      Organization = org_name,
      Agency = agency_name,
      AwardAmount = award_amt
    )
  })
  
  # Ensure numeric
  df <- df %>%
    mutate(AwardAmount = replace_na(AwardAmount, 0))
  
  cat("Cleaning complete. Saving", nrow(df), "records to", clean_file, "\n")
  write.csv(df, clean_file, row.names = FALSE)
  
  cat("\nSample of cleaned data:\n")
  print(head(df))
}

clean_nih_data()

File: scripts/day2/day2_scripts_Python/Day2_Application3_NIH.py
import os
import json
import pandas as pd

# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 2 - From raw output to usable data
# Application 3: Life Sciences - NIH RePORTER
# =============================================================================
# Goal: Clean and standardize raw JSON data into a usable tabular format.

def clean_nih_data():
    """
    Reads raw cached data from Day 1, cleans it, and saves it as a CSV.
    Demonstrates flattening nested dictionaries and date parsing.
    """
    raw_cache_file = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 
                                  'day1', 'data', 'cache', 'day1_app3_nih_raw.json')
    
    clean_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data', 'cache')
    os.makedirs(clean_dir, exist_ok=True)
    clean_file = os.path.join(clean_dir, 'day2_app3_nih_clean.csv')

    print(f"Loading raw data from {raw_cache_file}")
    with open(raw_cache_file, 'r', encoding='utf-8') as f:
        raw_data = json.load(f)
        
    records = raw_data.get('results', [])
    print(f"Loaded {len(records)} records. Starting cleaning process...")

    cleaned_records = []
    for record in records:
        # Extract basic info
        proj_num = record.get('project_num', '')
        title = record.get('project_title', '')
        
        # Extract nested info
        org = record.get('organization', {})
        org_name = org.get('org_name', 'Unknown') if org else 'Unknown'
        
        agency = record.get('agency_ic_admin', {})
        agency_name = agency.get('name', 'Unknown') if agency else 'Unknown'
        
        award_amount = record.get('award_amount', 0)
        
        cleaned_records.append({
            'ProjectNumber': proj_num,
            'Title': title,
            'Organization': org_name,
            'Agency': agency_name,
            'AwardAmount': award_amount
        })

    df = pd.DataFrame(cleaned_records)
    
    # Fill missing award amounts with 0 and convert to numeric
    df['AwardAmount'] = pd.to_numeric(df['AwardAmount']).fillna(0)
    
    print(f"Cleaning complete. Saving {len(df)} records to {clean_file}")
    df.to_csv(clean_file, index=False)
    
    print("\nSample of cleaned data:")
    print(df.head())

if __name__ == "__main__":
    clean_nih_data()
